compiler/rustc_smir/src/rustc_internal/mod.rs

```diff
@@ -24,14 +24,49 @@ use std::ops::Index;
 mod internal;
 pub mod pretty;
 
+/// Convert an internal Rust compiler item into its stable counterpart, if one exists.
+///
+/// # Warning
+///
+/// This function is unstable, and it's behavior may change at any point.
+/// E.g.: Items that were previously supported, may no longer be supported, or its translation may
+/// change.
+///
+/// # Panics
+///
+/// This function will panic if StableMIR has not been properly initialized.
 pub fn stable<'tcx, S: Stable<'tcx>>(item: S) -> S::T {
     with_tables(|tables| item.stable(tables))
 }
 
+/// Convert a stable item into its internal Rust compiler counterpart, if one exists.
+///
+/// # Warning
+///
+/// This function is unstable, and it's behavior may change at any point.
+/// Not every stable item can be converted to an internal one.
+/// Furthermore, items that were previously supported, may no longer be supported in newer versions.
+///
+/// # Panics
+///
+/// This function will panic if StableMIR has not been properly initialized.
 pub fn internal<'tcx, S: RustcInternal<'tcx>>(item: S) -> S::T {
     with_tables(|tables| item.internal(tables))
 }
 
+/// Retrieve the internal Rust compiler type context.
+///
+/// # Warning
+///
+/// This function is unstable, and it's behavior may change at any point.
+///
+/// # Panics
+///
+/// This function will panic if StableMIR has not been properly initialized.
+pub fn tcx<'tcx>() -> TyCtxt<'tcx> {
+    with_tables(|tables| tables.tcx)
+}
```

Comment:
These two functions are unsound. They trivially allow you to create a `TyCtxt<'static>` or a `Ty<'static>` and move it outside the `run!` invocation in safe code.

The only way I see them becoming safe is to use the `with_tables` scheme of passing a closure that then gets the item as an argument. This way we can have control over how long the item lives

Reply:
Indeed. I think I prefer closing this PR and changing the `internal()` interface to take tcx as an argument. What do you think?

Reply:
I removed `tcx()` and added the type context as an argument to `internal()` to avoid the issue you just mentioned.

Even though the new solution is a bit hacky, I think its safer and more usable than the `with_tables` solution. Let me know what you think.

Reply:
Oh that's even better!